Fix Java parser for annotated varargs by adding marker on `J.ArrayType`, as opposed to field on `J.VariableDeclarations`

knutwannheden opened the pull request:
Using a new `Varargs` marker.

TODO:
* [x] Fix all parsers
* [x] Phase out `varargs` field

- Fixes https://github.com/openrewrite/rewrite/issues/3881
- Fixes https://github.com/openrewrite/rewrite/issues/5742

knutwannheden committed:
Parse Java varargs annotations

Using a new `Varargs` marker.

knutwannheden committed:
Remove old code path

knutwannheden committed:
Simplify

knutwannheden committed:
Polish

timtebeek committed:
Merge branch 'main' into java-varargs-annotations

timtebeek committed:
Merge branch 'main' into java-varargs-annotations

timtebeek committed:
Replicate fix to the Java 25 parser

timtebeek committed:
Replicate fix to the Java 17 parser

timtebeek committed:
Replicate fix to the Java 11 parser

timtebeek committed:
Replicate fix to the Java 8 parser

timtebeek reviewed (approved):
Approved from my side; anything that had made you held off on merging?

I'd seen this fail in a number of projects, including when serializing the LST, so figured good to clear out.

timtebeek committed:
Merge branch 'main' into java-varargs-annotations

timtebeek committed:
Expect to go from two to one space before normalized `...`

timtebeek committed:
Mark the `J.VariableDeclarations.varargs` field as deprecated

timtebeek committed:
Merge branch 'main' into java-varargs-annotations

Commit:
Merge branch 'main' into java-varargs-annotations

sambsnyd reviewed (commented):
Basically all languages have some concept of varargs. Can we standardize this across all of them? Assuming it's agreed this is the way to go.

knutwannheden committed:
Merge branch 'main' into java-varargs-annotations

knutwannheden committed:
Drop `@ToBeRemoved` on deprecated `varargs` field

JS/TS, Python, and Go all use this field for their own native
varargs/star-args syntax (rest parameters, `*args`/`**kwargs`,
variadic types). Removing the field requires coordinated migration
across all five frontends, so a fixed removal date isn't realistic.
The `@Deprecated` annotation remains.

knutwannheden commented:
Thanks @greg-at-moderne for the survey — couple of corrections after digging into each.

**Scala** is implemented, but not using the `varargs` field for varargs. It repurposes the `Space` slot to store whitespace before the colon in field type ascription (`ScalaTreeVisitor.scala:2772`, e.g. `given IntSchema : SchemaFor[Int]`). Cleaning that up is independent of varargs and can be its own small PR (probably an `S.SpaceBeforeColon` marker on `J.VariableDeclarations`).

The other languages share a more interesting pattern: **JS/TS, Python, and Go all actively read/write this same field for their own native varargs syntax**, not just relay it over RPC:

- JS/TS prints `...args` rest parameters (`print.ts:497-499`)
- Python prints `*args`/`**kwargs` (`printer.py:1852-1854` — the `Space` is the whitespace before `*`)
- Go prints variadic `...T` (`go_printer.go:332-334`)

So this isn't a Java field that other languages happen to relay — it's a shared whitespace bucket every variable-declaration-having frontend is squatting on. Only C# is pure relay (it uses a `params` modifier natively).

That reshapes the plan. I think the right path is: land this PR as Java-only now, then track the cross-language work separately:

1. **Java** (this PR) — `Varargs` marker on `J.ArrayType`
2. **Groovy** — same pattern (also `String...` syntax)
3. **JS/TS** — marker on the rest-parameter representation
4. **Python** — dedicated field/marker for `*` / `**`
5. **Go** — marker on variadic-param representation
6. **Scala** — orthogonal: stop misusing the field for `:` whitespace

Once 1–5 land we can drop the field; until then it stays. I've also dropped the `@ToBeRemoved(after = "2026-04-01")` annotation in the latest push — no realistic removal date until the cross-language migration completes, and a stale date is worse than none. I'll open a tracking issue for 2–6.

knutwannheden commented on rewrite-java/src/test/java/org/openrewrite/java/format/SpacesVisitorTest.java:
```diff
@@ -373,7 +373,7 @@ void method(  String   first  ,   int  ...   values  ) {
             """
             class Test {
-                void method(String first, int... values) {}
+                void method(String first, int ... values) {}
```
Yep, intentional — `SpacesVisitor` doesn't have a rule that strips the space, just normalizes it. Going from two spaces to one is the expected behavior.

knutwannheden commented on rewrite-java/src/main/java/org/openrewrite/java/tree/J.java:
```diff
@@ -6109,6 +6109,8 @@ final class VariableDeclarations implements J, Statement, TypedTree {
         @Nullable
         @Getter
+        @Deprecated
+        @ToBeRemoved(after = "2026-04-01")
         Space varargs;
```
Updated based on the cross-language audit (see [main thread comment](https://github.com/openrewrite/rewrite/pull/6098#issuecomment-4307970752)): keeping `@Deprecated` (right signal for Java recipe authors — use the marker), but removed `@ToBeRemoved` since JS/TS, Python, and Go are all actively using the field for their own native varargs syntax. Removal needs the cross-language migration tracked above.

knutwannheden commented on rewrite-java/src/main/java/org/openrewrite/java/marker/Varargs.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.UUID;
+
+/**
+ * Indicates that a {@link org.openrewrite.java.tree.J.ArrayType} represents a varargs parameter type.
+ * When present, the array type should be printed using {@code ...} syntax instead of {@code []}.
+ * <p>
+ * For example, {@code String...} instead of {@code String[]}.
+ */
+@Value
+@With
+public class Varargs implements Marker {
+    UUID id;
```
Sam reviewed on Apr 1 — marking resolved.
